File: VideoMergeIOS/VideoMerger/Editor/Data/DraftStore.swift
//
//  DraftStore.swift
//  VideoMerger
//
//  iOS 移植自 Android DraftStore.kt
//  编辑器草稿存储 —— 单槽位自动保存（JSON），返回首页不丢内容
//

import Foundation

enum DraftStore {

    struct DraftInfo {
        let name: String
        let clipCount: Int
        let subtitleCount: Int
        let updatedAt: Double
    }

    static var draftURL: URL {
        let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("editor_draft", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("current.json")
    }

    static func save(_ project: EditorProject) {
        do {
            let data = try JSONEncoder().encode(project)
            try data.write(to: draftURL, options: .atomic)
        } catch {
            print("[DraftStore] 保存失败: \(error)")
        }
    }

    static func load() -> EditorProject? {
        guard let data = try? Data(contentsOf: draftURL) else { return nil }
        return try? JSONDecoder().decode(EditorProject.self, from: data)
    }

    static func clear() {
        try? FileManager.default.removeItem(at: draftURL)
    }

    /// 首页展示用轻量信息；无有效草稿返回 nil
    static func peek() -> DraftInfo? {
        guard let p = load() else { return nil }
        let clipCount = p.mainTrack?.clips.count ?? 0
        if clipCount == 0 && p.subtitles.isEmpty { return nil }
        return DraftInfo(name: p.name, clipCount: clipCount,
                         subtitleCount: p.subtitles.count, updatedAt: p.updatedAt)
    }
}

File: VideoMergeAndroid/app/src/main/java/com/moon/videomerger/editor/data/DraftStore.kt
package com.moon.videomerger.editor.data

import android.content.Context
import kotlinx.serialization.json.Json
import java.io.File

/**
 * 编辑器草稿存储 —— 单槽位自动保存。
 *
 * 文件：filesDir/editor_draft/current.json（项目 JSON，素材本体在 filesDir/editor_media）。
 * 返回首页不丢内容；首页提供「继续编辑」入口与删除操作。
 */
object DraftStore {

    private const val DIR = "editor_draft"
    private const val FILE = "current.json"

    /** 首页「继续编辑」卡片所需的轻量信息 */
    data class DraftInfo(
        val name: String,
        val clipCount: Int,
        val subtitleCount: Int,
        val updatedAt: Long,
    )

    private fun draftFile(context: Context): File =
        File(File(context.filesDir, DIR).apply { mkdirs() }, FILE)

    private val json = Json {
        ignoreUnknownKeys = true
    }

    /** 项目 → JSON（纯函数，便于单测锁定草稿格式） */
    internal fun encodeProject(project: EditorProject): String =
        json.encodeToString(EditorProject.serializer(), project)

    /** JSON → 项目；格式损坏返回 null */
    internal fun decodeProject(text: String): EditorProject? = try {
        json.decodeFromString(EditorProject.serializer(), text)
    } catch (_: Exception) {
        null
    }

    fun save(context: Context, project: EditorProject) {
        try {
            draftFile(context).writeText(encodeProject(project))
        } catch (_: Exception) {
            // 草稿保存失败不影响编辑流程
        }
    }

    fun load(context: Context): EditorProject? = try {
        val file = draftFile(context)
        if (file.exists()) decodeProject(file.readText()) else null
    } catch (_: Exception) {
        null
    }

    fun clear(context: Context) {
        try {
            draftFile(context).delete()
        } catch (_: Exception) {
        }
    }

    /** 首页展示用轻量信息（不完整反序列化 UI 状态）；无有效草稿返回 null */
    fun peek(context: Context): DraftInfo? {
        val project = load(context) ?: return null
        val clipCount = project.mainTrack?.clips?.size ?: 0
        if (clipCount == 0 && project.subtitles.isEmpty()) return null
        return DraftInfo(
            name = project.name,
            clipCount = clipCount,
            subtitleCount = project.subtitles.size,
            updatedAt = project.updatedAt,
        )
    }
}
